[FLINK-36066][runtime] Introducing the AdaptiveGraphManager component

noorall opened the pull request:
## What is the purpose of the change

To support the incremental generation of JobGraph and update the StreamGraph based on runtime information, we plan to introduce the AdaptiveGraphManager component, which will implement the AdaptiveGraphGenerator interface. The AdaptiveGraphGenerator interface is responsible for:

1. Generating the JobGraph
2. Responding to upstream job vertex finished events, generating JobVertex, and updating it to the JobGraph
3. Providing StreamGraphContext to allow StreamGraphOptimizer to add, delete, and modify StreamNode and StreamEdge


## Brief change log

- Implement the AdaptiveGraphGenerator interface
- Implement the StreamGraphContext interface
- Some UT cases

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

noorall commented:
@flinkbot run azure

noorall commented:
@flinkbot run azure

JunRuiLee reviewed (commented):
Thanks @noorall for your contribution. I've reviewed parts of the PR, except for the AdaptiveGraphManager and the tests. I'll submit comments for this part first. PTAL

noorall commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/util/StreamEdgeUpdateRequestInfo.java:
```diff
@@ -0,0 +1,62 @@
+    private StreamPartitioner<?> outputPartitioner;
+
+    public StreamEdgeUpdateRequestInfo(String edgeId, Integer sourceId, Integer targetId) {
+        this.edgeId = edgeId;
```
> I think sourceId and targetId are enough. The related changes can also be removed.

In a multi-input operator, it is possible for a source operator to have multiple edges pointing to the same multi-input operator. Therefore, it is not enough to identify a unique stream edge solely by sourceId and targetId.

JunRuiLee reviewed (commented):
Thank @noorall for updating and splitting the commits. I've left some comments; please take a look. Overall, I have a few suggestions:

1. Some logic needs to be refactored, and these areas should be covered by tests.
2. Methods with complex logic should have detailed and meaningful comments for explanation.
3. Some parts of the AdaptiveGraphManager are copied from StreamingJobGraphGenerator. I suggest refactoring StreamingJobGraphGenerator so that AdaptiveGraphManager can use these methods, reducing code redundancy and improving maintainability.

zhuzhurk reviewed (commented):
Thanks for creating this PR! @noorall 
I have left a few comments regarding the first commit.

JunRuiLee commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,1053 @@
+
+/** Default implementation for {@link AdaptiveGraphGenerator}. */
+@Internal
+public class AdaptiveGraphManager implements AdaptiveGraphGenerator {
```
When creating a `JobGraph` through this class, the `UserArtifact`, `UserJarBlobKeys`, and `Classpaths` from the `StreamGraph` are lost. 

Additionally, these cases lack test coverage.

noorall commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroupComputeUtil.java:
```diff
@@ -101,6 +105,96 @@ public static Map<JobVertexID, ForwardGroup> computeForwardGroups(
+        if (sourceForwardGroup.isMaxParallelismDecided()
+                && targetForwardGroup.isMaxParallelismDecided()
+                && sourceForwardGroup.getMaxParallelism()
+                        > targetForwardGroup.getMaxParallelism()) {
```
> Why use '>'? Could you add a comment to explain this?

If the max parallelism of the source is larger than that of the target, and the decided parallelism is equal to the max parallelism, the target can never achieve the same parallelism as the source, which may make the forward edge not working properly. Therefore, we need use '>'.

noorall commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/util/ImmutableStreamNode.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public @Nullable StreamOperatorFactory<?> getOperatorFactory() {
+        return streamNode.getOperatorFactory();
```
> In what case it can be null?

It depends on the `streamNode.getOperatorFactory()` which returns a `@Nullable` value.

JunRuiLee commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphContext.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Modifies stream edges within the StreamGraph.
+     *
+     * @param requestInfos the stream edges to be modified.
```
```
/**
 * Atomically modifies stream edges within the StreamGraph.
 *
 * <p>This method ensures that all the requested modifications to stream edges are 
 * applied atomically. This means that if any modification fails, none of the 
 * changes will be applied, maintaining the consistency of the StreamGraph.
 *
 * @param requestInfos the stream edges to be modified.
 * @return true if all modifications were successful and applied atomically, false otherwise.
 */
```

JunRuiLee commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java:
```diff
@@ -18,2596 +19,380 @@
 /** Tests for {@link StreamingJobGraphGenerator}. */
 @SuppressWarnings("serial")
-class StreamingJobGraphGeneratorTest {
+class StreamingJobGraphGeneratorTest extends JobGraphGeneratorTestBase {
```
I think that the incremental tests need to cover more aspects than those for the StreamingJobGraphGenerator. They should cover the incremental parts in more detail. Therefore, all tests for the StreamingJobGraphGenerator should be in the JobGraphGeneratorTestBase.

JunRuiLee commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManagerTest.java:
```diff
@@ -0,0 +1,274 @@
+                                    }
+                                });
+
+        result.addSink(
```
the SinkFunction is deprecated, we should use the SinkV2:

```
result.sinkTo(new DiscardingSink<>());
```

JunRuiLee commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,672 @@
+            ClassLoader userClassloader,
+            StreamGraph streamGraph,
+            Executor serializationExecutor,
+            @Nullable JobID jobId) {
```
I think the job id should be obtained from the StreamGraph instead of being passed as a parameter.

JunRuiLee commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,672 @@
+                    transitiveNonChainableOutEdgesMap
+                            .get(startNodeId)
+                            .forEach(
+                                    streamEdge -> {
+                                        this.startAndEndNodeIdToForwardGroupMap.put(
+                                                streamEdge.getSourceId(), forwardGroup);
```
streamEdge -> this.startAndEndNodeIdToForwardGroupMap.put(
                                            streamEdge.getSourceId(), forwardGroup)

JunRuiLee reviewed (commented):
Thank @noorall for the update. Such a complex refactoring is no easy task. I've left some comments, please take a look

noorall commented:
> Thank @noorall for the update. Such a complex refactoring is no easy task. I've left some comments, please take a look

Thanks for your review. The PR has been updated according to your comments.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/DefaultStreamGraphContext.java:
```diff
@@ -0,0 +1,222 @@
+            return false;
+        }
+
+        if (!ForwardGroupComputeUtil.canTargetMergeIntoSourceForwardGroup(
```
Looks to it it invokes `canTargetMergeIntoSourceForwardGroup(...)` twice. One here and one in `mergeForwardGroup(...)`.

noorall commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/util/ImmutableStreamNode.java:
```diff
@@ -0,0 +1,70 @@
+        return streamNode.getId();
+    }
+
+    public @Nullable StreamOperatorFactory<?> getOperatorFactory() {
```
> The returned `operatorFactory` can be modifiable. In which case Flink needs to get an `operatorFactory` from an `immutableStreamNode`?

In the adaptive broadcast join and adaptive skewd join optimization strategies, flink need to get it (to be precise, AdaptiveJoinOperatorFactory) to get some necessary optimization information.

zhuzhurk replied:
`AdaptiveBroadcastJoin` will modify the `AdaptiveJoinOperatorFactory` in some cases. It may be improper if the `AdaptiveJoinOperatorFactory` is retrieved using this method, because here it is an `ImmutableStreamNode`.

noorall replied:
> `AdaptiveBroadcastJoin` will modify the `AdaptiveJoinOperatorFactory` in some cases. It may be improper if the `AdaptiveJoinOperatorFactory` is retrieved using this method, because here it is an `ImmutableStreamNode`.

Agree with you! I have moved this method to StreamGraphContext.

noorall commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/forwardgroup/StreamNodeForwardGroup.java:
```diff
@@ -0,0 +1,197 @@
+                            || maxParallelism >= parallelism);
+        }
+
+        if (this.isParallelismDecided() || this.isMaxParallelismDecided()) {
```
> It's better to introduce two methods `updateNodeParallelism` and `updateNodeMaxParallelism` and invoke them only when any value really changes. Besides that, looks to me here the invocation of `setMaxParallelism(int)` will always set the `parallelismConfigured` of a `StreamNode` to true. Not sure if it is expected? cc @JunRuiLee

Do you mean `setParallelism(int)`? The invocation of `setMaxParallelism(int)` wouldn't change the `parallelismConfigured`. And it seems expected to me, based on the following code: https://github.com/apache/flink/blob/77e7ac076c35d229fab89b8b6ae15138b0e41a22/flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java#L908C49-L909C95

zhuzhurk replied:
Yes I meant `setParallelism(int)`.
In the code from the shared link, it can work because the parallelism will be unset for a StreamNode without user specified parallelism. (https://github.com/apache/flink/blob/77e7ac076c35d229fab89b8b6ae15138b0e41a22/flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java#L856)

We can do it here because the parallelism is also unset for any StreamNode without user specified parallelism, before doing such forward group computing?

noorall commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroupComputeUtilTest.java:
```diff
@@ -59,6 +68,28 @@ void testIsolatedVertices() throws Exception {
+        // Different from the job vertex forward group, the stream node forward group is allowed to
+        // contain only one single stream node, as these groups may merge with other groups in the
+        // future.
+        checkGroupSize(groups, 3, 1);
```
> B.T.W. will it create a job vertex forward group if the size is 1 when the job vertex is created?

No, it won’t. A forward group containing only one job vertex is meaningless.

noorall commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/DefaultStreamGraphContext.java:
```diff
@@ -0,0 +1,222 @@
+
+        if (frozenNodeToStartNodeMap.containsKey(targetNodeId)) {
+            LOG.info(
+                    "Modification for edge {} is not allowed as the target node with id {} is in frozen list.",
```
> Will this kind of logs happen too frequently?

No, it won’t happen too frequently. The first log only happens when attempting to modify a reused edge, and the second log only happens when an exception occurs in the optimization strategy (usually, when the optimization strategy is executed, the downstream job vertex is not created).

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -1655,6 +1703,21 @@ && arePartitionerAndExchangeModeChainable(
     }
 
+    public static boolean isSourceChainable(StreamNode streamNode, StreamGraph streamGraph) {
+        if (!streamNode.getInEdges().isEmpty()
```
Why requiring the inputs to be empty?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -673,18 +601,32 @@ private void setChaining() {
+
+            // Progressive hash generation is required in adaptive graph generator.
+            if (hashGenerator != null) {
+                hashGenerator.accept(currentNodeId);
```
From the perspective of this method, it is not a generator because it generates nothing.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -731,13 +677,20 @@ private List<StreamEdge> createChain(
-                                nonChainable.getTargetId(),
-                                (k) -> chainInfo.newChain(nonChainable.getTargetId())),
-                        chainEntryPoints);
+                if (canCreateNewChain) {
```
Comments are needed to explain what's it for.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -777,18 +730,28 @@ private List<StreamEdge> createChain(
-                            : new StreamConfig(new Configuration());
+            StreamConfig config;
+            if (currentNodeId.equals(startNodeId)) {
+                JobVertex jobVertex = jobVertexBuildContext.getJobVertex(startNodeId);
+                config =
+                        jobVertex != null
+                                ? new StreamConfig(jobVertex.getConfiguration())
+                                : new StreamConfig(
+                                        createJobVertex(
+                                                        chainInfo,
+                                                        serializationExecutor,
+                                                        jobVertexBuildContext)
+                                                .getConfiguration());
```
Maybe
```
                JobVertex jobVertex = jobVertexBuildContext.getJobVertex(startNodeId);
                if (jobVertex == null) {
                    jobVertex = createJobVertex( chainInfo, serializationExecutor, jobVertexBuildContext);
                }
                config = new StreamConfig(jobVertex.getConfiguration());
```

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+
+    private final List<Map<Integer, byte[]>> legacyHashes;
+
+    // Records the id of stream node which job vertex is created.
```
-> Records the ids of stream nodes of which the job vertices are already created

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -1655,6 +1703,21 @@ && arePartitionerAndExchangeModeChainable(
         return true;
     }
 
+    public static boolean isSourceChainable(StreamNode streamNode, StreamGraph streamGraph) {
```
-> isChainableSource

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+                transitiveNonChainableOutEdgesMap, topologicallySortedChainedStreamNodesMap);
+    }
+
+    private List<StreamNode> getEnterPoints(
```
IIUC, this is the actual `getChainEntryPoints()`? And the existing `getChainEntryPoints()` is `getChainEntryPointsReadyToCreateChain()`?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+        return enterPoints;
+    }
+
+    private void traverseFullGraph(
```
The method name is not explanatory. Could you add comments to explain what this method does?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+                enterPoints.add(sourceNode);
+            }
+        }
+        return enterPoints;
```
How about to let this method simply return `chainedStreamNodesMap`? In this way we do not need to create and maintain the `enterPoints`. And the `chainedStreamNodesMap` can be created within this method instead of being a parameter. Also , maybe the name can be `createChainEntryPoints()`?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+                    topologicallySortedChainedStreamNodesMap,
+                    visitedStartNodes);
+        }
+        if (currentNodeId.equals(startNodeId)) {
```
Why only skipping start nodes? And why adding it to the visited nodes in the end? Looks to me this may result in multiple visitations to one node.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+                .computeIfAbsent(startNodeId, k -> new ArrayList<>())
+                .addAll(nonChainableOutputs);
+
+        topologicallySortedChainedStreamNodesMap.get(startNode).add(currentNode);
```
Looks to me it is more clear to move this operation to the beginning of this method, after the visited-check?

noorall commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroupComputeUtilTest.java:
```diff
@@ -182,7 +319,56 @@ private static Set<ForwardGroup> computeForwardGroups(JobVertex... vertices) {
+                                                return ((StreamNodeForwardGroup) group).size();
+                                            }
+                                        })
+                                .collect(Collectors.toList()))
```
> Looks to me this change is not needed if invoking this method as proposed in other comments.

The size method in ForwardGroup is only used in tests, so we removed it when abstracting the interface.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+                                        startNodeId, opIntermediateOutputs, jobVertexBuildContext));
+    }
+
+    private void setVertexNonChainedOutputsConfig(
```
It's better to explain in method level comments about the difference from that in `StreamJobGraphGenerator`.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+    private final Map<Integer, Map<StreamEdge, NonChainedOutput>> intermediateOutputsCaches;
+
+    // Records the id of the stream node that produces the IntermediateDataSet.
+    private final Map<IntermediateDataSetID, Integer> intermediateDataSetIdToProducerMap;
```
To confirm, is it just used for testing? Or it will be used in production code in later PRs?

noorall replied:
> To confirm, is it just used for testing? Or it will be used in production code in later PRs?

Will be used in later PRs.

zhuzhurk commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManagerTest.java:
```diff
@@ -0,0 +1,354 @@
+        StreamGraph streamGraph = env.getStreamGraph();
+        streamGraph.setDynamic(true);
+        JobGraph jobGraph = createJobGraph(streamGraph);
+        assertThat(jobGraph.getVerticesSortedTopologicallyFromSources().size()).isEqualTo(4);
```
Only one source can be chained in this case? input3 will not be included in the multi-input chain?

noorall replied:
> Only one source can be chained in this case? input3 will not be included in the multi-input chain?

Yes, the parallelism of input2 and input3 is 1, and only the parallelism of input1 meets the conditions of sourceChain.

noorall commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -0,0 +1,697 @@
+        StreamNode startNode = streamGraph.getStreamNode(startNodeId);
+        for (StreamEdge inEdges : startNode.getInEdges()) {
+            Integer sourceNodeId = inEdges.getSourceId();
+            if (!hashes.containsKey(sourceNodeId)) {
```
> In which case it can happen that the hash of the `sourceNode` has been created, if that the `sourceNode` has finished or it is a chained source?

Currently, this happens in both cases.

noorall replied:
> When is the hash of a chained source created?

Currently this only happens while the JobGraph is initializing. In `buildAndGetChainEntryPoints`, if it is a chainable source, a hash will be generated for it immediately.

noorall replied:
> In which case it can happen that the hash of the `sourceNode` has been created, if that the `sourceNode` has finished or it is a chained source?

It seems the checks are redundant; it only needs to verify whether the upstream hash has been generated.

noorall commented:
> Thanks for addressing all the comments above. @noorall I have a few comments regarding the first 2 commits.

Thanks for your review. The PR has been updated according to your comments.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -1678,10 +1678,27 @@ public static ResultPartitionType determineUndefinedResultPartitionType(
-        return downStreamVertex.getInEdges().size() == 1 && isChainableInput(edge, streamGraph);
+        return downStreamVertex.getInEdges().size() == 1
+                && isChainableInput(edge, streamGraph)
+                && checkAndSetParallelismForChainableNodes(edge, streamGraph);
```
It's a bit weird to update node parallelism/maxParallelism in a method named `isChainable()`.

zhuzhurk replied:
How about to do the update when doing `ForwardPartitioner` conversion?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java:
```diff
@@ -227,10 +236,21 @@ public Integer getProducerStreamNodeId(IntermediateDataSetID intermediateDataSet
+     * @param intermediateDataSetID the ID of the IntermediateDataSet
+     * @return the stream edges that subscribe to the IntermediateDataSet
+     */
+    public List<StreamEdge> getOutputStreamEdges(IntermediateDataSetID intermediateDataSetID) {
```
intermediateDataSetID -> intermediateDataSetId

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/DefaultStreamGraphContext.java:
```diff
@@ -206,6 +204,36 @@ private void modifyOutputPartitioner(
+                } else {
+                    targetEdge.setPartitioner(
+                            ((ForwardForConsecutiveHashPartitioner<?>) targetEdge.getPartitioner())
+                                    .getHashPartitioner());
```
Unlike `RescalePartitioner`,  `HashPartitioner` will be worse than `ForwardPartitioner`. So a `ForwardPartitioner` should be used in this case only if the nodes have different parallelisms (the forward group merge fails).

noorall commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/DefaultStreamGraphContext.java:
```diff
@@ -166,33 +181,16 @@ private void modifyOutputPartitioner(
-                targetEdge.setPartitioner(new RescalePartitioner<>());
-            }
-        }
+        tryConvertForwardPartitionerAndMergeForwardGroup(targetEdge);
```
if (targetEdge.getPartitioner() instanceof ForwardPartitioner) {
            tryConvertForwardPartitionerAndMergeForwardGroup(targetEdge);
        }

zhuzhurk reviewed (approved):
Thanks for addressing all the comments! @noorall
LGTM.

JunRuiLee reviewed (approved):
LGTM, and I also ran the CI with this PR as the default job graph generation mode for batch jobs. Everything seems fine
